components/scream/ekat/cmake/EkatMpiUtils.cmake

```diff
@@ -1,28 +1,29 @@
 # Detect the library that provides MPI
+set (EKAT_CMAKE_DIR ${CMAKE_CURRENT_LIST_DIR})
 macro (GetMpiDistributionName DISTRO_NAME)
   execute_process(COMMAND ${CMAKE_CXX_COMPILER} 
-                  ${CMAKE_CURRENT_SOURCE_DIR}/cmake/TryCompileOpenMPI.cpp
+                  ${EKAT_CMAKE_DIR}/TryCompileOpenMPI.cpp
                   OUTPUT_QUIET ERROR_QUIET
                   RESULT_VARIABLE HAVE_OPENMPI_MPI)
 
   if (HAVE_OPENMPI_MPI)
-    set (DISTRO_NAME "openmpi")
+    set (${DISTRO_NAME} "openmpi")
   else ()
     execute_process(COMMAND ${CMAKE_CXX_COMPILER} 
                     ${CMAKE_CURRENT_SOURCE_DIR}/cmake/TryCompileMPICH.cpp
                     OUTPUT_QUIET ERROR_QUIET
                     RESULT_VARIABLE HAVE_MPICH_MPI)
 
     if (HAVE_MPICH_MPI)
-      set (DISTRO_NAME "mpich")
+      set (${DISTRO_NAME} "mpich")
```

Comment:
Just curious: does the ${...} here mean that DISTRO_NAME isn't a regular string variable in the context of the `set` command? Why not just `set(DISTRO_NAME "unknown")`?

Reply:
DISTRO_NAME contains the name of the variable used to store the result. It is, in a way, a double level of indirection.

CMake does not have a way to _return_ values from functions or macros. When you pass a variable to a function/macro, cmake interpret it as a string. Focusing on macros (function add an extra scope layer, but are otherwise similar), if you have `macro(foo my_var)` and try to do `set(my_var "blah")`, it will work only if, by chance, the variable is called `my_var` also outside the macro. The correct way to write in the variable is to get the content of `my_var` and interpret it as _the name of the output variable_, that is `set (${my_var} "blah")`.

I wish cmake had a simpler way of returning form macros/functions. This can be complicated. Btw, I had it wrong at the beginning (I had what you suggested), but it _happened_ to work, cause at the call site the variable was also called DISTRO_NAME, which worked. When I accidentally changed the var name at the call site to `MPI_DISTRIBUTION_NAME`, everything broke, which helped me track down this cmake subtle issue.

Reply:
Gross. Thanks for the explanation!

Reply:
Gross indeed!